From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] wm2: Remove vif state entry for deleted interface

--- a/core/src/wm2/src/wm2_radio.c
+++ b/core/src/wm2/src/wm2_radio.c
@@ -959,6 +959,7 @@ wm2_vconf_recalc(const char *ifname, bool force)
     bool dpp_enabled;
     bool want;
     bool has;
+    int del_vif = 0;
 
     LOGD("%s: recalculating", ifname);
 
@@ -976,6 +976,9 @@ wm2_vconf_recalc(const char *ifname, bool force)
                                        &vstate)))
         wm2_vstate_init(&vstate, ifname);
 
+    if (!want)
+        del_vif = 1;
+
     if (want == true) {
         wm2_l2uf_if_enable(ifname);
         if (strcmp(vconf.mode, "sta") == 0)
@@ -1154,6 +1154,13 @@ wm2_vconf_recalc(const char *ifname, bool force)
         WARN_ON(!ovsdb_table_upsert(&table_Wifi_VIF_State, &vstate, false));
     }
 
+    if (del_vif) {
+        ovsdb_table_delete_simple(&table_Wifi_VIF_State,
+                                  SCHEMA_COLUMN(Wifi_VIF_State, if_name),
+                                  vconf.if_name);
+        LOGD("%s: VIF state column deleted", vconf.if_name);
+    }
+
     wm2_delayed_recalc_cancel(wm2_vconf_recalc, ifname);
 }
 
