From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Center frequency fix

--- a/core/src/lib/common/inc/util.h
+++ b/core/src/lib/common/inc/util.h
@@ -150,5 +150,5 @@ char  *file_get(const char *path);
 const int *unii_5g_chan2list(int chan, int width);
 const int *unii_6g_chan2list(int chan, int width);
 bool is_private_ip(char *ip_str);
-
+int unii_5g_centerfreq(const char *ht_mode, const int channel);
 #endif /* UTIL_H_INCLUDED */
--- a/core/src/lib/common/src/util.c
+++ b/core/src/lib/common/src/util.c
@@ -1435,3 +1435,21 @@ bool ascii2hex(const char *input, char *
 {
     return bin2hex((unsigned char *)input, strlen(input), output, size) == 0;
 }
+
+int unii_5g_centerfreq(const char *ht_mode, const int channel)
+{
+    const int width = atoi(strlen(ht_mode) > 2 ? ht_mode + 2 : "20");
+    const int *chans = unii_5g_chan2list(channel, width);
+    int sum = 0;
+    int cnt = 0;
+
+    if (!chans)
+        return 0;
+
+    while (*chans) {
+        sum += *chans;
+        cnt++;
+        chans++;
+    }
+    return sum / cnt;
+}
--- a/core/src/lib/hostap/src/hapd.c
+++ b/core/src/lib/hostap/src/hapd.c
@@ -723,25 +723,10 @@ hapd_util_get_oper_chwidth(const struct schem
 static int
 hapd_util_get_oper_centr_freq_idx(const struct schema_Wifi_Radio_Config *rconf)
 {
-    const int width = atoi(strlen(rconf->ht_mode) > 2 ? rconf->ht_mode + 2 : "20");
-    const int *chans;
-    int sum = 0;
-    int cnt = 0;
-
-    if (!strcmp(rconf->freq_band, SCHEMA_CONSTS_RADIO_TYPE_STR_6G))
-        chans = unii_6g_chan2list(rconf->channel, width);
-    else
-        chans = unii_5g_chan2list(rconf->channel, width);
-
-    if (WARN_ON(!chans))
+    if (!rconf->ht_mode_exists || !rconf->channel_exists)
         return 0;
 
-    while (*chans) {
-        sum += *chans;
-        cnt++;
-        chans++;
-    }
-    return sum / cnt;
+    return unii_5g_centerfreq(rconf->ht_mode, rconf->channel);
 }
 
 int
