From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Fetch vif state rrm value from 
hostapd conf

--- a/core/src/lib/hostap/src/hapd.c
+++ b/core/src/lib/hostap/src/hapd.c
@@ -1333,6 +1333,9 @@ hapd_bss_get(struct hapd *hapd,
         vstate->ft_psk = atoi(p);
     if ((vstate->btm_exists = (p = ini_geta(conf, "bss_transition"))))
         vstate->btm = atoi(p);
+    if ((vstate->rrm_exists = (p = ini_geta(conf, "rrm_neighbor_report"))))
+        vstate->rrm = atoi(p);
+
 
     SCHEMA_SET_STR(vstate->multi_ap, hapd_map_int2str(atoi(map ?: "0")));
     SCHEMA_SET_INT(vstate->dpp_cc, atoi(ini_geta(conf, "dpp_configurator_connectivity") ?: "0"));
