From: Shubham Pawar <p.shubham@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Set beacon interval

--- a/core/src/lib/hostap/src/hapd.c
+++ b/core/src/lib/hostap/src/hapd.c
@@ -1079,6 +1079,9 @@ hapd_conf_gen(struct hapd *hapd,
 
     if (vconf->group_rekey_exists && vconf->group_rekey >= 0)
         csnprintf(&buf, &len, "wpa_group_rekey=%d\n", vconf->group_rekey);
+    
+    if (rconf->bcn_int_exists)
+        csnprintf(&buf, &len, "beacon_int=%d\n", rconf->bcn_int);
 
     if (vconf->wpa_exists)
         return hapd_conf_gen_security(hapd, buf, &len, vconf);
