From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap: Enable hostapd config option to use 
driver interface address

--- a/core/src/lib/hostap/src/hapd.c
+++ b/core/src/lib/hostap/src/hapd.c
@@ -1089,6 +1089,7 @@ hapd_conf_gen(struct hapd *hapd,
     csnprintf(&buf, &len, "ieee80211ac=%d\n", hapd_11ac_enabled(rconf));
     csnprintf(&buf, &len, "%s", hapd->ieee80211ax ? "" : "#");
     csnprintf(&buf, &len, "ieee80211ax=%d\n", hapd_11ax_enabled(rconf));
+    csnprintf(&buf, &len, "use_driver_iface_addr=1\n");
     csnprintf(&buf, &len, "%s", vconf->dpp_cc_exists ? "" : "#");
     csnprintf(&buf, &len, "dpp_configurator_connectivity=%d\n", vconf->dpp_cc ? 1 : 0);
 
