From: aparna.salian <aparna.salian@globaledgesoft.com>
Subject: [PATCH] lib/hostap:  Override AP VIF's channel with 
connected bhaul station's channel

--- a/core/src/lib/hostap/inc/opensync-hapd.h
+++ b/core/src/lib/hostap/inc/opensync-hapd.h
@@ -45,6 +45,7 @@ struct hapd {
     char htcaps[256];
     char vhtcaps[512];
     int noscan; /* skip initial obss scan */
+    int ext_chan_override;
     void (*sta_connected)(struct hapd *hapd, const char *mac, const char *keyid);
     void (*sta_disconnected)(struct hapd *hapd, const char *mac);
     void (*ap_enabled)(struct hapd *hapd);
--- a/core/src/lib/hostap/src/hapd.c
+++ b/core/src/lib/hostap/src/hapd.c
@@ -688,7 +688,7 @@ hapd_11ax_enabled(const struct schema_Wi
 }
 
 static const char *
-hapd_ht_caps(const struct schema_Wifi_Radio_Config *rconf)
+hapd_ht_caps(const struct schema_Wifi_Radio_Config *rconf, struct hapd *hapd)
 {
     if (!rconf->ht_mode_exists) return "";

@@ -703,7 +703,7 @@ hapd_ht_caps(const struct schema_Wifi_Ra
     }
 
     if (!strcmp(rconf->freq_band, SCHEMA_CONSTS_RADIO_TYPE_STR_2G)) {
-        switch (rconf->channel) {
+        switch (hapd->ext_chan_override ? hapd->ext_chan_override : rconf->channel) {
             case 0:
                 return "[HT40+] [HT40-]";
             case 1 ... 7:
@@ -782,12 +782,13 @@ hapd_util_get_oper_chwidth(const struct schem
 }
 
 static int
-hapd_util_get_oper_centr_freq_idx(const struct schema_Wifi_Radio_Config *rconf)
+hapd_util_get_oper_centr_freq_idx(const struct schema_Wifi_Radio_Config *rconf, struct hapd *hapd)
 {
     if (!rconf->ht_mode_exists || !rconf->channel_exists)
         return 0;
 
-    return unii_5g_centerfreq(rconf->ht_mode, rconf->channel);
+    return unii_5g_centerfreq(rconf->ht_mode,
+                hapd->ext_chan_override ? hapd->ext_chan_override : rconf->channel);
 }
 
 int
@@ -1072,7 +1073,8 @@ hapd_conf_gen(struct hapd *hapd,
     csnprintf(&buf, &len, "logger_syslog_level=3\n");
     csnprintf(&buf, &len, "ssid=%s\n", vconf->ssid);
     csnprintf(&buf, &len, "hw_mode=%s\n", hapd_band2hwmode(rconf->freq_band));
-    csnprintf(&buf, &len, "channel=%d\n", rconf->channel);
+    csnprintf(&buf, &len, "channel=%d\n",
+            hapd->ext_chan_override ? hapd->ext_chan_override : rconf->channel);
     csnprintf(&buf, &len, "ignore_broadcast_ssid=%d\n", closed);
     csnprintf(&buf, &len, "wmm_enabled=1\n");
     csnprintf(&buf, &len, "%s", hapd->respect_multi_ap ? "" : "#");
@@ -1097,7 +1097,7 @@ hapd_conf_gen(struct hapd *hapd,
         if (hapd->ieee80211ax && hapd_11ax_enabled(rconf)) {
             csnprintf(&buf, &len, "he_oper_chwidth=%d\n", hapd_util_get_oper_chwidth(rconf));
             csnprintf(&buf, &len, "he_oper_centr_freq_seg0_idx=%d\n",
-                      hapd_util_get_oper_centr_freq_idx(rconf));
+                      hapd_util_get_oper_centr_freq_idx(rconf, hapd));
         }
         if (hapd->ieee80211ac && hapd_11ac_enabled(rconf) &&
             (strcmp(rconf->freq_band, SCHEMA_CONSTS_RADIO_TYPE_STR_6G))) {
@@ -1103,7 +1103,7 @@ hapd_conf_gen(struct hapd *hapd,
             (strcmp(rconf->freq_band, SCHEMA_CONSTS_RADIO_TYPE_STR_6G))) {
             csnprintf(&buf, &len, "vht_oper_chwidth=%d\n", hapd_util_get_oper_chwidth(rconf));
             csnprintf(&buf, &len, "vht_oper_centr_freq_seg0_idx=%d\n",
-                      hapd_util_get_oper_centr_freq_idx(rconf));
+                      hapd_util_get_oper_centr_freq_idx(rconf, hapd));
         }
     }
 
@@ -1114,7 +1114,7 @@ hapd_conf_gen(struct hapd *hapd,
 
     if (hapd->ieee80211n)
         csnprintf(&buf, &len, "ht_capab=%s %s\n",
-                  hapd->htcaps, hapd_ht_caps(rconf));
+                  hapd->htcaps, hapd_ht_caps(rconf, hapd));
 
     if (hapd->ieee80211ac)
         csnprintf(&buf, &len, "vht_capab=%s\n", hapd->vhtcaps);
