--- a/package/network/services/ppp/Makefile
+++ b/package/network/services/ppp/Makefile
@@ -222,9 +222,13 @@ define Package/ppp/install
 	$(INSTALL_DIR) $(1)/usr/sbin
 	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/sbin/pppd $(1)/usr/sbin/
 	$(INSTALL_DIR) $(1)/etc/ppp
+ifndef CONFIG_PACKAGE_opensync
 	$(INSTALL_CONF) ./files/etc/ppp/chap-secrets $(1)/etc/ppp/
+endif
 	$(INSTALL_DATA) ./files/etc/ppp/filter $(1)/etc/ppp/
+ifndef CONFIG_PACKAGE_opensync
 	$(INSTALL_DATA) ./files/etc/ppp/options $(1)/etc/ppp/
+endif
 	$(LN) /tmp/resolv.conf.ppp $(1)/etc/ppp/resolv.conf
	$(INSTALL_DIR) $(1)/lib/netifd/proto
	$(INSTALL_BIN) ./files/ppp.sh $(1)/lib/netifd/proto/
